haystack/utils/base_serialization.py

```diff
@@ -52,3 +53,93 @@ class does not have a `from_dict` method.
 
     return obj_class.from_dict(data["data"])
+
+
+def serialize_value(value: Any) -> Any:
+    """
+    Serializes a value into a format suitable for storage or transmission.
+
+    Handles various types including:
+    - Haystack dataclass objects (Answer, Document, etc.)
+    - Primitive types (returned as is)
+    - Lists of primitives (returned as is)
+    - Lists of complex types (recursively serialized)
+    - Dictionaries (recursively serialized)
+
+    :param value: The value to serialize
+    :returns: The serialized representation of the value
+    """
+
+    if hasattr(value, "to_dict") and callable(getattr(value, "to_dict")):
+        serialized_value = value.to_dict()
+        serialized_value["_type"] = value.__class__.__name__
+        return serialized_value
+
+    # this is a hack to serialize inputs that don't have a to_dict
+    elif hasattr(value, "__dict__"):
+        return {"_type": value.__class__.__name__, "attributes": value.__dict__}
+
+    # recursively serialize all inputs in a dict
+    elif isinstance(value, dict):
+        return {k: serialize_value(v) for k, v in value.items()}
+
+    # recursively serialize all inputs in lists or tuples
+    elif isinstance(value, list):
+        return [serialize_value(item) for item in value]
+
+    return value
+
+
+# pylint: disable=too-many-return-statements
+def deserialize_value(value: Any) -> Any:
+    """
+    Deserializes a value from its serialized representation.
+
+    Handles various types including:
+    - Haystack dataclass objects (Answer, Document, etc.)
+    - Primitive types (returned as is)
+    - Lists of primitives (returned as is)
+    - Lists of complex types (recursively deserialized)
+    - Dictionaries (recursively deserialized)
+
+    :param value: The serialized value to deserialize
+    :returns: The deserialized value
+    """
+
+    # None or primitive types are returned as is
+    if not value or isinstance(value, (str, int, float, bool)):
+        return value
+
+    # list of primitive types are returned as is
+    if isinstance(value, list) and all(isinstance(i, (str, int, float, bool)) for i in value):
+        return value
+
+    if isinstance(value, list):
+        # list of lists are called recursively
+        if all(isinstance(i, list) for i in value):
+            return [deserialize_value(i) for i in value]
+        # list of dicts are called recursively
+        if all(isinstance(i, dict) for i in value):
+            return [deserialize_value(i) for i in value]
+
+    # Define the mapping of types to their deserialization functions
+    _type_deserializers = {
+        "Answer": Answer.from_dict,
+        "ChatMessage": ChatMessage.from_dict,
+        "Document": Document.from_dict,
+        "ExtractedAnswer": ExtractedAnswer.from_dict,
+        "GeneratedAnswer": GeneratedAnswer.from_dict,
+        "SparseEmbedding": SparseEmbedding.from_dict,
+    }
```

Comment:
Instead of using a predefined set of deserializers couldn't we follow a similar methodology like we do for `component_from_dict` where we import the `_type` field, check if the imported class has a `from_dict` attribute and if so use that? That way we wouldn't need to create a list of hard-coded deserializers.